File: apps/cli/src/output.ts
/**
 * JSON-first output helpers for the `frick` CLI.
 *
 * Default output: JSON Lines (one JSON object per logical record) on stdout.
 * `--pretty` (or `--json=pretty`) switches to indented JSON on stdout.
 * Errors always go to stderr as a single JSON object.
 *
 * The CLI deliberately does not have human-only output modes — the spec
 * principle is "machine-readable first, human summaries layered on top of
 * the stable JSON shape." Pretty mode is still JSON, just indented.
 */

export type OutputMode = "json" | "pretty";

export interface OutputOptions {
  mode: OutputMode;
  stdout: NodeJS.WritableStream;
  stderr: NodeJS.WritableStream;
}

export function resolveOutputMode(flags: Record<string, string | boolean | undefined>): OutputMode {
  if (flags.pretty === true) return "pretty";
  const json = flags.json;
  if (typeof json === "string" && json === "pretty") return "pretty";
  return "json";
}

export function emit(record: unknown, opts: OutputOptions): void {
  const payload =
    opts.mode === "pretty" ? `${JSON.stringify(record, null, 2)}\n` : `${JSON.stringify(record)}\n`;
  opts.stdout.write(payload);
}

export function emitError(
  error: { code: string; message: string; details?: unknown },
  opts: OutputOptions,
): void {
  const body = { error };
  const payload =
    opts.mode === "pretty" ? `${JSON.stringify(body, null, 2)}\n` : `${JSON.stringify(body)}\n`;
  opts.stderr.write(payload);
}

File: crates/frick-cli/src/output.rs
//! JSON-first output helpers (ported from `apps/cli/src/output.ts`).
//!
//! Default output is JSON Lines (one compact JSON object per record) on
//! stdout. `--pretty` / `--json=pretty` switches to 2-space-indented JSON.
//! Errors always go to stderr as a single `{ "error": { … } }` object.

use std::io::Write;

use serde_json::Value;

use crate::argv::ParsedArgs;
use crate::errors::CliError;

/// Output mode (`OutputMode` in TS).
#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub enum OutputMode {
    /// Compact JSON Lines (default).
    Json,
    /// 2-space-indented JSON.
    Pretty,
}

/// Output sinks + the resolved mode. Streams are `&mut dyn Write` so tests can
/// capture into buffers and the binary passes stdout/stderr.
pub struct Output<'a> {
    /// Resolved output mode.
    pub mode: OutputMode,
    /// Stdout sink (records).
    pub stdout: &'a mut dyn Write,
    /// Stderr sink (errors).
    pub stderr: &'a mut dyn Write,
}

/// `resolveOutputMode` — `--pretty` (boolean true) or `--json=pretty` ⇒ pretty.
#[must_use]
pub fn resolve_output_mode(parsed: &ParsedArgs) -> OutputMode {
    if parsed.flag_bool_present("pretty") {
        return OutputMode::Pretty;
    }
    if parsed.flag_str("json") == Some("pretty") {
        return OutputMode::Pretty;
    }
    OutputMode::Json
}

impl Output<'_> {
    /// Emit one record to stdout, newline-terminated.
    pub fn emit(&mut self, record: &Value) {
        let payload = match self.mode {
            OutputMode::Pretty => serde_json::to_string_pretty(record),
            OutputMode::Json => serde_json::to_string(record),
        }
        .unwrap_or_else(|_| "null".to_string());
        let _ = writeln!(self.stdout, "{payload}");
    }

    /// Emit a structured error to stderr (`{ "error": { … } }`).
    pub fn emit_error(&mut self, error: &CliError) {
        let body = error.to_body();
        let payload = match self.mode {
            OutputMode::Pretty => serde_json::to_string_pretty(&body),
            OutputMode::Json => serde_json::to_string(&body),
        }
        .unwrap_or_else(|_| "null".to_string());
        let _ = writeln!(self.stderr, "{payload}");
    }
}
